### PR TITLE
Apl 807 validators verification

### PR DESCRIPTION
**Pull request by nemezxxx**

### CODE REVIEW

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/model/DexOrder.java [hunk 1]
```diff
@@ -14,6 +14,7 @@
 import lombok.EqualsAndHashCode;
 
 import java.math.BigDecimal;
+import lombok.Getter;
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/model/DexOrder.java]: Unused import

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/model/UserEthDepositInfo.java [hunk 2]
```diff
@@ -5,6 +5,8 @@
 import lombok.NoArgsConstructor;
 
 import java.math.BigDecimal;
+import lombok.Getter;
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/model/UserEthDepositInfo.java]: Unused import

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/service/DexValidationServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+        
+        String hisAddress = null;
+        if (hisOrder.getType() == OrderType.BUY) 
+            hisAddress = hisOrder.getFromAddress();
+        else hisAddress = hisOrder.getToAddress();
+        log.debug("selected: {}",hisAddress);
+        // here we have double conversion, gw-eth-wei        
+        Long averageGasPriceGw = ethGasInfo.getAverageSpeedPrice();  
+        BigInteger hisEthBalanceWei = getEthBalanceWei(/*hisOffer.getToAddress(),*/hisAddress, hisOrder.getPairCurrency());
+        BigDecimal averageGasPriceEth = EthUtil.gweiToEth(averageGasPriceGw);
```
[the remark this answers is not in the document]
Comment (AndrewBoyarsky) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/service/DexValidationServiceImpl.java]: No, why do you not use direct gwei to wei conversion?  Just add one method to EthUtil.

[the remark this answers is not in the document]
Comment (nemezxxx) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/service/DexValidationServiceImpl.java]: @AndrewBoyarsky  
it has been already done. You remark is taken into consideration. In fact, this piece looks like this now
https://prnt.sc/p5k4bs
This remark is a little bit outdated

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/model/DexOrder.java (hunk 1 above)
Comment (nemezxxx) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/model/DexOrder.java]: Removed it, thanx

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/model/UserEthDepositInfo.java (hunk 2 above)
Comment (nemezxxx) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/model/UserEthDepositInfo.java]: The same, done

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/service/DexValidationServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+        
+        // 1) Checking out whether HE has the corresponding amount on his APL balance        
+        Long hisAccountID = hisOrder.getAccountId();
+        log.debug("hisAccountID(apl): {}, his fromAddr : {}, his toAddr: {}", hisAccountID,hisOrder.getFromAddress(),hisOrder.getToAddress());
+        Long hisUnconfirmedAplBalance = getAplUnconfirmedBalance(hisAccountID);
+        Long hisAplBalance = getAplBalanceAtm(hisAccountID);  
+        Long balanceDelta = hisAplBalance - hisUnconfirmedAplBalance;
+        boolean isFrozenEnough = balanceDelta >= hisOrder.getOrderAmount();
+        log.debug("isFrozenEnough: {} ", isFrozenEnough);
+        if (!isFrozenEnough) return OFFER_VALIDATE_ERROR_APL_FREEZE;
```
[the remark this answers is not in the document]
Comment (nemezxxx) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/service/DexValidationServiceImpl.java]: does not have to. I have to validate my counteroffer. In this situation, I shouldn't care about somebody else's matching to my order. In fact, I have to validate whether my interlocutor is able to pay. For this purpose, I have a contract and two orders and nothing else. 
All the other matchings should be rejected with further putting their initial orders into their "pending" case. 
That's the way the cookie crumbles

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/service/DexMatcherServiceImpl.java [hunk 5]
```diff
@@ -60,49 +64,69 @@ public void deinitialize() {
...
-            case APL: return validateOfferAPL(offer);
-            case ETH: return validateOfferETH(offer);
-            case PAX: return validateOfferPAX(offer);
-            default: return false;
+
+            case ETH: { 
+                // return validateOfferETH(myOffer,hisOffer);
+                if (myOrder.getType() == OrderType.SELL) 
+                    return validateOfferSellAplEth(myOrder, hisOrder); 
+                else return validateOfferBuyAplEth(myOrder, hisOrder);
```
[the remark this answers is not in the document]
Comment (nemezxxx) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/service/DexMatcherServiceImpl.java]: done

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/service/DexMatcherServiceImpl.java [hunk 6]
```diff
@@ -60,49 +64,69 @@ public void deinitialize() {
...
+                // return validateOfferETH(myOffer,hisOffer);
+                if (myOrder.getType() == OrderType.SELL) 
+                    return validateOfferSellAplEth(myOrder, hisOrder); 
+                else return validateOfferBuyAplEth(myOrder, hisOrder);                
+            }
+            
+            case PAX: {
+                if (myOrder.getType() == OrderType.SELL) 
+                    return validateOfferSellAplPax(myOrder, hisOrder); 
+                else return validateOfferBuyAplPax(myOrder, hisOrder);
```
[the remark this answers is not in the document]
Comment (nemezxxx) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/service/DexMatcherServiceImpl.java]: done

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/service/DexOrderProcessor.java [hunk 7]
```diff
@@ -153,7 +160,74 @@ private void broadcastWhenConfirmed(Transaction txToBroadcast, Transaction uncon
...
+                }
+                break;
+            }
+            
+            default: // return OFFER_VALIDATE_ERROR_IN_PARAMETER;
+                rx = OFFER_VALIDATE_ERROR_IN_PARAMETER;
+                break;
+        }        
+        
+        log.debug("validation result: {}",rx);
```
Comment (leonidstr) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/exchange/service/DexOrderProcessor.java]: add order id to the log

**Pull request merged**